### PR TITLE
Properties conditional directive and macro parser in IdMap

### PR DESCRIPTION
**Pull request by Justsnoopy30**
Adds a parser for conditional directives (#if, #else, #endif) and macros (MC_VERSION) used for block.properties, item.properties, and entity.properties in IdMap.

### CODE REVIEW

**Review by coderbot16 — CHANGES_REQUESTED** [review]
A few notes from an initial review:

* It looks like this includes an update of Gradle & Loom, these would preferably be kept to a separate PR
* It looks like this depends on #169, so I will need to review and approve that first before fully reviewing / approving this PR
* The preprocessor code should probably be moved to a separate file instead of being placed in IdMap
* Have you verified that nested conditions work with this approach?

**Review by coderbot16 — CHANGES_REQUESTED** [review]
There are still a lot of changes to `gradlew` and `gradlew.bat` - could these be reverted please?

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 1]
```diff
@@ -19,6 +17,9 @@
 import net.minecraft.util.Identifier;
 import net.minecraft.util.InvalidIdentifierException;
 
+import static java.lang.Integer.parseInt;
+import static net.coderbot.iris.shaderpack.PropertiesPreprocessor.parseProperties;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:21]: Avoid using `import static` where possible, prefer referencing those methods directly.

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 2]
```diff
@@ -34,16 +35,16 @@
...
 
 	/**
-	 * a map that contains render layers for blocks in block.properties
+	 * A map that contains render layers for blocks in block.properties
 	 */
-	private Map<Identifier, RenderLayer> blockRenderLayerMap = Maps.newHashMap();
+	private Map<Identifier, RenderLayer> blockRenderLayerMap = new HashMap<>();
 
-	IdMap(Path shaderPath) {
+	public IdMap(Path shaderPath) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:47]: Why does this need to be public?

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 3]
```diff
@@ -59,20 +60,47 @@
 		// TODO: Properly override block render layers
 	}
 
+	/**
+	 * Loads properties from a properties file in a shaderpack path
+	 */
 	private static Optional<Properties> loadProperties(Path shaderPath, String name) {
-		Properties properties = new Properties();
+		if (shaderPath == null) return Optional.empty();
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:67]: `if` statements should always be surrounded with braces. Also, why is this null check here in the first place? `shaderPath` should never be null.

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 4]
```diff
@@ -59,20 +60,47 @@
...
+
+		String fileContents = readProperties(shaderPath, name);
+		if (fileContents == null) return Optional.empty();
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:70]: Should have braces here too

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 5]
```diff
@@ -99,7 +127,7 @@
 			int intId;
 
 			try {
-				intId = Integer.parseInt(key.substring(keyPrefix.length()));
+				intId = parseInt(key.substring(keyPrefix.length()));
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:130]:
```suggestion
				intId = Integer.parseInt(key.substring(keyPrefix.length()));
```

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 6]
```diff
@@ -59,20 +60,47 @@
...
 			return Optional.empty();
 		}
 
 		return Optional.of(properties);
 	}
 
+	private static String readProperties(Path shaderPath, String name) {
+		try {
+			return new String(Files.readAllBytes(shaderPath.resolve(name)), StandardCharsets.UTF_8);
+		} catch (IOException e) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:97]: It's possible that this will be something other than a `NoSuchFileException`, that should be handled separately

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 7]
```diff
@@ -59,20 +60,47 @@
...
+		String fileContents = readProperties(shaderPath, name);
+		if (fileContents == null) return Optional.empty();
 
+		List<String> lines = parseProperties(name, fileContents);
+
+		StringBuilder processed = new StringBuilder();
+		for (String line : lines) {
+			processed.append(line);
+			processed.append('\n');
+		}
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:78]: Would probably be best to move this to PropertiesPreprocessor and have parseProperties return a string.

File: src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static net.coderbot.iris.gl.shader.StandardMacros.getMcVersion;
+
+public class PropertiesPreprocessor {
+	private static final ImmutableMap<String, String> MACRO_CONSTANTS = ImmutableMap.of("MC_VERSION", getMcVersion());
+	
+	public static List<String> parseProperties(String fileName, String fileContents) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java:15]: This should probably be named something like `process` or similar.

File: src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
+package net.coderbot.iris.shaderpack;
+
+import com.google.common.collect.ImmutableMap;
+import net.coderbot.iris.Iris;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static net.coderbot.iris.gl.shader.StandardMacros.getMcVersion;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java:10]: Avoid using `import static` due to code clarity issues

File: src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+							break;
+						default: {
+							Iris.logger.error("Invalid operator " + operator + " in " + fileName);
+							continue;
+						}
+					}
+				} else {
+					Iris.logger.warn("Unknown variable fileName " + variable + " in " + fileName);
+					currentlyParsingConditionalProperties = false;
+				}
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java:104]: Should probably have a separate function for parsing the expression within the `#if` directive and returning whether or not it passed, then that would make it easier to support `#elif`.

File: src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+import static net.coderbot.iris.gl.shader.StandardMacros.getMcVersion;
+
+public class PropertiesPreprocessor {
+	private static final ImmutableMap<String, String> MACRO_CONSTANTS = ImmutableMap.of("MC_VERSION", getMcVersion());
+	
+	public static List<String> parseProperties(String fileName, String fileContents) {
+		List<String> lines = new ArrayList<>();
+
+		boolean currentlyParsingConditionalProperties = false;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java:18]: Should have a more descriptive name for this... perhaps `currentConditionPassed` or something?

File: src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+		String currentConditional = null;
+
+		for (String line : fileContents.split("\\R")) {
+			String trimmedLine = line.trim();
+
+			if (trimmedLine.startsWith("#if ")) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PropertiesPreprocessor.java:24]: Since nested directives are not supported, that should cause a hard error somewhere for now instead of breaking silently

**Review by coderbot16 — CHANGES_REQUESTED** [review]
A good start, noticed a few issues

File: src/main/java/net/coderbot/iris/shaderpack/IdMap.java [hunk 13]
```diff
@@ -34,16 +35,16 @@
...
 	 */
-	private Map<Identifier, Integer> blockPropertiesMap = Maps.newHashMap();
+	private Map<Identifier, Integer> blockPropertiesMap = new HashMap<>();
 
 	/**
-	 * a map that contains render layers for blocks in block.properties
+	 * A map that contains render layers for blocks in block.properties
 	 */
-	private Map<Identifier, RenderLayer> blockRenderLayerMap = Maps.newHashMap();
+	private Map<Identifier, RenderLayer> blockRenderLayerMap = new HashMap<>();
```
Comment (Kroppeb) [inline src/main/java/net/coderbot/iris/shaderpack/IdMap.java:45]: Why these changes?

Comment (Justsnoopy30) [reply src/main/java/net/coderbot/iris/shaderpack/IdMap.java:45]: Maps.newHashMap() is deprecated and should be replaced with new HashMap<>();

**Comment by coderbot16** [comment]
I noticed that a few comments I made weren't fully addressed - I've resolved the ones that have been addressed, but there are a few that have not been addressed.

In particular, I'm not sure how `shaderPath` would end up being null, if this is the case then crashing is probably a better option.

**Comment by Justsnoopy30** [comment]
> I noticed that a few comments I made weren't fully addressed - I've resolved the ones that have been addressed, but there are a few that have not been addressed, could you please review them?
> 
> In particular, I'm not sure how `shaderPath` would end up being null, if this is the case then crashing is probably a better option.

The shaderPath null check was a left-over from moving this from starline to trunk where completely disabling shaders (which can cause shaderPath to be null) isn't a thing. I'll remove that and address the other comments I missed tomorrow since I'm already in bed right now.

**Comment by coderbot16** [comment]
Sounds good, thanks.

**Review by coderbot16 — APPROVED** [review]
Looks good, thanks! Now just waiting on a merge of the other PR.

**Comment by Justsnoopy30** [comment]
> Looks good, thanks! Now just waiting on a merge of the other PR.

Do you still want me to create another method to parse the if statement or is it fine as-is?

**Comment by coderbot16** [comment]
I'm willing to merge this PR as is, but that would definitely be nice as well.

**Comment by coderbot16** [comment]
I squash merged the other PR, so it seems like this PR needs to be rebased on the current trunk, or the conflicts otherwise resolved.

**Review by coderbot16 — APPROVED** [review]
Thanks!
